ni_measurementlink_service/_internal/parameter/_message.py

```diff
@@ -0,0 +1,94 @@
+import struct
+
+from google.protobuf.internal import encoder, wire_format
+
+
+def _inner_message_encoder(field_index):
+    """Mimics google.protobuf.internal.MessageEncoder.
+
+    See EncodeField:
+    https://github.com/protocolbuffers/protobuf/blob/0b817d46d4ca1977d3dccf2442aeee3c9e98e3a1/python/google/protobuf/internal/encoder.py#L765C15-L765C15
+    """
+    tag = encoder.TagBytes(field_index, wire_format.WIRETYPE_LENGTH_DELIMITED)
+    encode_varint = _varint_encoder()
+
+    def _encode_message(write, value, deterministic):
+        write(tag)
+        bytes = value.SerializeToString()
+        encode_varint(write, len(bytes), deterministic)
+        write(bytes)
+
+    return _encode_message
+
+
+def _varint_encoder():
+    """Return an encoder for a basic varint value (does not include tag).
+
+    From google.protobuf.internal.encoder.py _VarintEncoder
+    """
+    local_int2byte = struct.Struct(">B").pack
+
+    def encode_varint(write, value, unused_deterministic=None):
+        bits = value & 0x7F
+        value >>= 7
+        while value:
+            write(local_int2byte(0x80 | bits))
+            bits = value & 0x7F
+            value >>= 7
+        return write(local_int2byte(bits))
+
+    return encode_varint
+
+
+def _inner_message_decoder(field_index, is_repeated, is_packed, key, new_default):
+    """Based on google.protobuf.internal.MessageDecoder.
```

Comment:
Is the point of this function to call `ParseFromString` instead of `_InternalParse`? That would be relevant information to document.

Ditto for the encoder and `_InternalSerialize`.

I would still like to know why this is the case. Is it due to upb vs. the old pure-Python parser? If you set the environment variable to use the old pure-Python parser, does it define `_InternalParse` and `_InternalSerialize`?

Reply:
Added comments to the method.